pkg/syncer/metadatasyncer.go

```diff
@@ -398,38 +339,68 @@ func InitMetadataSyncer(ctx context.Context, clusterFlavor cnstypes.CnsClusterFl
 	}
 
 	if metadataSyncer.clusterFlavor == cnstypes.CnsClusterFlavorGuest {
-		// Check the replicated FSS Configmap every 2 minutes
-		// When workload-domain-isolation FSS is enabled in csi-feature-states config-map, restart the container
-		if commonco.ContainerOrchestratorUtility.IsPVCSIFSSEnabled(ctx, common.WorkloadDomainIsolationFSS) &&
-			!commonco.ContainerOrchestratorUtility.IsCNSCSIFSSEnabled(ctx, common.WorkloadDomainIsolationFSS) {
+		// If workload-domain-isolation FSS is not enabled on guest cluster, then check the capabilities CR in
+		// supervisor cluster every 2 mins to check if there is a change in capability value from false to true.
+		// If so, restart the CSI container.
+		// NOTE: We can add other capabilities here when similar functionality is required. For
+		// workload-isolation-domain feature we are restarting the container when capability changes dynamically from
+		// false to true, but for other features instead of restarting CSI container, if possible we can implement
+		// init() function and it can initialize required things when capability value changes from false to true.
+		if !commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.WorkloadDomainIsolationFSS) {
 			go func() {
+				// If supervisor is old but TKR is new, it could happen that Capabilities CR is not
+				// present on the supervisor. Check if Capabilities CRD is registered on supervisor
+				// before starting a ticker to check for capability value from Capabilities CR.
+				restClientConfig := k8s.GetRestClientConfigForSupervisor(ctx,
+					metadataSyncer.configInfo.Cfg.GC.Endpoint, metadataSyncer.configInfo.Cfg.GC.Port)
+				apiextensionsClientSet, err := apiextensionsclientset.NewForConfig(restClientConfig)
+				if err != nil {
+					log.Errorf("failed to create apiextension clientset using config. Err: %+v", err)
+					return
+				}
+				_, err = apiextensionsClientSet.ApiextensionsV1().CustomResourceDefinitions().Get(ctx,
+					"capabilities.iaas.vmware.com", metav1.GetOptions{})
+				if err != nil {
+					if apierrors.IsNotFound(err) {
```

Comment:
if not found then i think we should wait for supervisor to get upgraded and once we find CRs are registered and capabilities are available, upon supervisor capability enablement, guest cluster container needs to be restarted.